File: THCS/2022_2023/BinhDinh_2223/Bai1/SIMILAR.cpp
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <vector>
#include <algorithm>
using namespace std;

// Hàm tính tổng các chữ số của một số n
int tong_cac_chu_so(int n) {
    int tong = 0;
    while (n > 0) {
        tong += n % 10;
        n /= 10;
    }
    return tong;
}

int main() {
    // Mở file đầu vào
    ifstream infile("SIMILAR.INP");
    int l, r;
    infile >> l >> r;
    infile.close();

    // Từ điển lưu danh sách các số theo tổng chữ số
    unordered_map<int, vector<int>> tu_dien;

    // Duyệt qua tất cả các số từ l đến r
    for (int n = l; n <= r; ++n) {
        int tong_chu_so = tong_cac_chu_so(n);  // Tính tổng các chữ số
        tu_dien[tong_chu_so].push_back(n);     // Thêm số vào danh sách tương ứng
    }

    // Tìm hiệu lớn nhất
    int hieu_lon_nhat = 0;
    for (auto &entry : tu_dien) {
        vector<int> &ds = entry.second;
        if (ds.size() > 1) {  // Nếu có nhiều hơn 1 số có cùng tổng chữ số
            int max_val = *max_element(ds.begin(), ds.end());
            int min_val = *min_element(ds.begin(), ds.end());
            hieu_lon_nhat = max(hieu_lon_nhat, max_val - min_val);
        }
    }

    // Ghi kết quả ra file đầu ra
    ofstream outfile("SIMILAR.OUT");
    outfile << hieu_lon_nhat << endl;
    outfile.close();

    return 0;
}

File: THCS/2022_2023/BinhDinh_2223/Bai1/SIMILAR.py
# Hàm tính tổng các chữ số của một số
def tong_cac_chu_so(n):
    return sum(int(c) for c in str(n))

# Đọc dữ liệu từ file SIMILAR.INP
with open("SIMILAR.INP", "r") as infile:
    l, r = map(int, infile.readline().split())

# Tạo một từ điển để lưu danh sách các số theo tổng chữ số
tu_dien = {}

# Duyệt qua tất cả các số trong đoạn từ l đến r
for n in range(l, r + 1):
    tong_chu_so = tong_cac_chu_so(n)  # Tính tổng các chữ số
    if tong_chu_so not in tu_dien:
        tu_dien[tong_chu_so] = []  # Tạo danh sách mới nếu chưa có
    tu_dien[tong_chu_so].append(n)  # Thêm số vào danh sách

# Tìm hiệu lớn nhất
hieu_lon_nhat = 0
for ds in tu_dien.values():
    if len(ds) > 1:  # Chỉ xét nếu có ít nhất 2 số có cùng tổng chữ số
        hieu_lon_nhat = max(hieu_lon_nhat, max(ds) - min(ds))

# Ghi kết quả ra file SIMILAR.OUT
with open("SIMILAR.OUT", "w") as outfile:
    outfile.write(str(hieu_lon_nhat) + "\n")
